# Mastering Pandas: Unveiling Movie Trends with IMDB Data

---

## Table of Contents & Learning Objectives

In this notebook, we'll dive deep into **Pandas**, a powerful Python library for data manipulation and analysis. Our goal is to extract meaningful insights from a real-world dataset – IMDB movie data – to understand movie trends and solve practical business problems.

### Learning Objectives:
By the end of this session, you will be able to:
- Load and perform initial exploration of datasets.
- Merge multiple datasets based on common keys.
- Understand and apply the `.apply()` method for custom data transformations.
- Master the `.groupby()` method for advanced data aggregation, filtering, and transformation.
- Solve real-world data analysis problems using Pandas.

### Content Overview:
- **Introduction to IMDB Use Case**
    - Merging `movies` & `directors` datasets
    - Initial IMDB data exploration
- **`apply()` Method for Row/Column-wise Operations**
- **`groupby()` for Advanced Data Analysis**
    - Group-based Aggregation
    - Group-based Filtering
    - Group-based Apply

---

## **Case Study: Unveiling Movie Industry Trends with IMDB Data**

### Business Problem:
Imagine you are a Data Scientist at a leading entertainment analytics firm. Your client, a major film studio, wants to understand key trends in the movie industry to make informed decisions about future productions. They are particularly interested in:
- Identifying successful director profiles.
- Understanding the financial performance of movies.
- Discovering patterns in movie releases over time.

### Our Data Source:
We'll be working with data from **IMDB (Internet Movie Database)**, a comprehensive online database of information related to films. Specifically, we have two datasets:
- `movies.csv`: Contains details about individual movies, such as budget, revenue, popularity, and release date.
- `directors.csv`: Provides information about the directors, including their name and gender.

Let's download and read the IMDB dataset files into our Colab environment.

In [ ]:
# Install the gdown library to download files from Google Drive
!pip install --upgrade gdown

  Attempting uninstall: gdown
    Found existing installation: gdown 5.2.2
    Uninstalling gdown-5.2.2:
      Successfully uninstalled gdown-5.2.2


In [ ]:
# Download the movies.csv file from Google Drive
!gdown 1s2TkjSpzNc4SyxqRrQleZyDIHlc7bxnd

Downloading...
From: https://drive.google.com/uc?id=1s2TkjSpzNc4SyxqRrQleZyDIHlc7bxnd
To: /content/movies.csv
100% 112k/112k [00:00<00:00, 84.9MB/s]


### Understanding `gdown`

Before loading our data, we used `gdown`. This is a Python package that allows us to download files directly from Google Drive into our Colab environment. The `!pip install` command ensures that `gdown` is installed, and then `!gdown` with the specific Google Drive file ID downloads the `movies.csv` and `directors.csv` files to our current working directory. This is a common practice when your datasets are hosted on Google Drive.

In [ ]:
# Download the directors.csv file from Google Drive
!gdown 1Ws-_s1fHZ9nHfGLVUQurbHDvStePlEJm

Downloading...
From: https://drive.google.com/uc?id=1Ws-_s1fHZ9nHfGLVUQurbHDvStePlEJm
To: /content/directors.csv
100% 65.4k/65.4k [00:00<00:00, 84.4MB/s]


We have successfully downloaded two CSV files:
- `movies.csv`: Contains detailed information for each movie.
- `directors.csv`: Contains information about the directors, which we will later link to the movies dataset.

Now, let's load these into Pandas DataFrames and begin our initial exploration.

In [ ]:
# Import the pandas library for data manipulation, aliasing it as 'pd'
import pandas as pd
# Import the numpy library for numerical operations, aliasing it as 'np'
import numpy as np

### Importing Essential Libraries

We begin by importing `pandas` as `pd` and `numpy` as `np`. `pandas` is the cornerstone of data manipulation in Python, providing DataFrames for structured data. `numpy` is crucial for numerical operations and is often used in conjunction with `pandas`.

In [ ]:
movies = pd.read_csv('/content/movies.csv')
# Display the first 5 rows of the DataFrame to get a quick overview
print("\n--- Movies DataFrame (First 5 Rows) ---")
display(movies.head())

# Get a concise summary of the DataFrame, including data types and non-null values
print("\n--- Movies DataFrame Info ---")
movies.info()

# Generate descriptive statistics of the DataFrame's numerical columns
print("\n--- Movies DataFrame Descriptive Statistics ---")
display(movies.describe())

# Check for missing values in each column
print("\n--- Movies DataFrame Missing Values ---")
display(movies.isnull().sum())


--- Movies DataFrame (First 5 Rows) ---


,Unnamed: 0,id,budget,popularity,revenue,title,vote_average,vote_count,director_id,year,month,day
0,0,43597,237000000,150,2787965087,Avatar,7.2,11800,4762,2009,Dec,Thursday
1,1,43598,300000000,139,961000000,Pirates of the Caribbean: At World's End,6.9,4500,4763,2007,May,Saturday
2,2,43599,245000000,107,880674609,Spectre,6.3,4466,4764,2015,Oct,Monday
3,3,43600,250000000,112,1084939099,The Dark Knight Rises,7.6,9106,4765,2012,Jul,Monday
4,5,43602,258000000,115,890871626,Spider-Man 3,5.9,3576,4767,2007,May,Tuesday



--- Movies DataFrame Info ---
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1465 entries, 0 to 1464
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Unnamed: 0    1465 non-null   int64  
 1   id            1465 non-null   int64  
 2   budget        1465 non-null   int64  
 3   popularity    1465 non-null   int64  
 4   revenue       1465 non-null   int64  
 5   title         1465 non-null   object 
 6   vote_average  1465 non-null   float64
 7   vote_count    1465 non-null   int64  
 8   director_id   1465 non-null   int64  
 9   year          1465 non-null   int64  
 10  month         1465 non-null   object 
 11  day           1465 non-null   object 
dtypes: float64(1), int64(8), object(3)
memory usage: 137.5+ KB

--- Movies DataFrame Descriptive Statistics ---


,Unnamed: 0,id,budget,popularity,revenue,vote_average,vote_count,director_id,year
count,1465.000000,1465.000000,1.465000e+03,1465.000000,1.465000e+03,1465.000000,1465.000000,1465.000000,1465.000000
mean,1627.391809,45225.191126,4.802295e+07,30.855973,1.432539e+08,6.368191,1146.396587,5040.192491,2002.615017
std,1187.182894,1189.096396,4.935541e+07,34.845214,2.064918e+08,0.818033,1578.077438,258.059631,8.680141
min,0.000000,43597.000000,0.000000e+00,0.000000,0.000000e+00,3.000000,1.000000,4762.000000,1976.000000
25%,639.000000,44236.000000,1.400000e+07,11.000000,1.738013e+07,5.900000,216.000000,4845.000000,1998.000000
50%,1425.000000,45022.000000,3.300000e+07,23.000000,7.578164e+07,6.400000,571.000000,4964.000000,2004.000000
75%,2393.000000,45990.000000,6.600000e+07,41.000000,1.792469e+08,6.900000,1387.000000,5179.000000,2009.000000
max,4768.000000,48395.000000,3.800000e+08,724.000000,2.787965e+09,8.300000,13752.000000,6204.000000,2016.000000



--- Movies DataFrame Missing Values ---


,0
Unnamed: 0,0
id,0
budget,0
popularity,0
revenue,0
title,0
vote_average,0
vote_count,0
director_id,0
year,0


In [ ]:
654/1465

0.44641638225255975

In [ ]:
display(movies.describe(include="all"))

,Unnamed: 0,id,budget,popularity,revenue,title,vote_average,vote_count,director_id,year,month,day
count,1465.000000,1465.000000,1.465000e+03,1465.000000,1.465000e+03,1465,1465.000000,1465.000000,1465.000000,1465.000000,1465,1465
unique,NaN,NaN,NaN,NaN,NaN,1465,NaN,NaN,NaN,NaN,12,7
top,NaN,NaN,NaN,NaN,NaN,El Mariachi,NaN,NaN,NaN,NaN,Dec,Friday
freq,NaN,NaN,NaN,NaN,NaN,1,NaN,NaN,NaN,NaN,193,654
mean,1627.391809,45225.191126,4.802295e+07,30.855973,1.432539e+08,NaN,6.368191,1146.396587,5040.192491,2002.615017,NaN,NaN
std,1187.182894,1189.096396,4.935541e+07,34.845214,2.064918e+08,NaN,0.818033,1578.077438,258.059631,8.680141,NaN,NaN
min,0.000000,43597.000000,0.000000e+00,0.000000,0.000000e+00,NaN,3.000000,1.000000,4762.000000,1976.000000,NaN,NaN
25%,639.000000,44236.000000,1.400000e+07,11.000000,1.738013e+07,NaN,5.900000,216.000000,4845.000000,1998.000000,NaN,NaN
50%,1425.000000,45022.000000,3.300000e+07,23.000000,7.578164e+07,NaN,6.400000,571.000000,4964.000000,2004.000000,NaN,NaN
75%,2393.000000,45990.000000,6.600000e+07,41.000000,1.792469e+08,NaN,6.900000,1387.000000,5179.000000,2009.000000,NaN,NaN


In [ ]:
# 1. Read the directors.csv
df_directors = pd.read_csv('/content/directors.csv')
display(df_directors.head())

# 2. Get a concise summary of the DataFrame, including data types and non-null values
print("\n--- Directors DataFrame Info ---")
df_directors.info()

# 2. Tell me the descriptive stats for the directors file
print("\n--- Directors DataFrame Descriptive Statistics ---")
display(df_directors.describe(include="all"))

# 3. Give me the number of null values for the directors file
print("\n--- Directors DataFrame Missing Values ---")
display(df_directors.isnull().sum())

,Unnamed: 0,director_name,id,gender
0,0,James Cameron,4762,Male
1,1,Gore Verbinski,4763,Male
2,2,Sam Mendes,4764,Male
3,3,Christopher Nolan,4765,Male
4,4,Andrew Stanton,4766,Male



--- Directors DataFrame Info ---
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2349 entries, 0 to 2348
Data columns (total 4 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   Unnamed: 0     2349 non-null   int64 
 1   director_name  2349 non-null   object
 2   id             2349 non-null   int64 
 3   gender         1724 non-null   object
dtypes: int64(2), object(2)
memory usage: 73.5+ KB

--- Directors DataFrame Descriptive Statistics ---


,Unnamed: 0,director_name,id,gender
count,2349.000000,2349,2349.000000,1724
unique,NaN,2349,NaN,2
top,NaN,Brian Herzlinger,NaN,Male
freq,NaN,1,NaN,1574
mean,1174.000000,NaN,5936.000000,NaN
std,678.242213,NaN,678.242213,NaN
min,0.000000,NaN,4762.000000,NaN
25%,587.000000,NaN,5349.000000,NaN
50%,1174.000000,NaN,5936.000000,NaN
75%,1761.000000,NaN,6523.000000,NaN



--- Directors DataFrame Missing Values ---


,0
Unnamed: 0,0
director_name,0
id,0
gender,625


## Initial Data Exploration: What Can We Discover?

Before diving into complex analysis, let's think about the kind of insights we can gather from this raw `movies` dataset:

-   **Top Movies**: We can identify the most popular movies using the `popularity` column.
-   **Highest Rated Films**: By looking at `vote_average`, we can find movies that audiences loved.
-   **Release Patterns**: The `year`, `month`, and `day` columns allow us to see how many movies are released each year or even by month.
-   **Budget Analysis**: We can find movies with the highest production `budget` and analyze their financial performance.

### Beyond the Obvious: Asking Deeper Questions

However, to truly understand the film industry, we need to ask more nuanced questions that often require combining and manipulating data:

-   **Most Productive Directors**: Which directors have made the most films?
-   **High-Budget Directors**: Which directors are typically associated with large-scale, high-budget productions?
-   **Monthly Performance Trends**: Can we find the highest and lowest-rated movies released in a particular month or year?

To answer these deeper questions, we'll need to learn how to effectively combine our datasets and use powerful Pandas features like `apply()` and `groupby()`.

### Addressing the 'Unnamed: 0' Column

Take a closer look at the `movies` DataFrame's head. You'll notice a column named **`Unnamed: 0`**. This column appears to be an artifact, likely representing the original index from the CSV file that Pandas read as a regular column.

**Challenge**: How can we instruct Pandas to treat this first column as the DataFrame's index rather than a regular data column, effectively "getting rid" of the `Unnamed: 0` column?

In [ ]:
movies = pd.read_csv('movies.csv', index_col=0) # Use the first column as the index
# Display the updated DataFrame to confirm the change
display(movies.head())

,id,budget,popularity,revenue,title,vote_average,vote_count,director_id,year,month,day
0,43597,237000000,150,2787965087,Avatar,7.2,11800,4762,2009,Dec,Thursday
1,43598,300000000,139,961000000,Pirates of the Caribbean: At World's End,6.9,4500,4763,2007,May,Saturday
2,43599,245000000,107,880674609,Spectre,6.3,4466,4764,2015,Oct,Monday
3,43600,250000000,112,1084939099,The Dark Knight Rises,7.6,9106,4765,2012,Jul,Monday
5,43602,258000000,115,890871626,Spider-Man 3,5.9,3576,4767,2007,May,Tuesday


In [ ]:
movies = pd.read_csv('movies.csv')
movies.drop(["Unnamed: 0"], axis=1, inplace=True) # axis parameter : drop the column (1) or drop the row (0)

In [ ]:
movies.head()

,id,budget,popularity,revenue,title,vote_average,vote_count,director_id,year,month,day
0,43597,237000000,150,2787965087,Avatar,7.2,11800,4762,2009,Dec,Thursday
1,43598,300000000,139,961000000,Pirates of the Caribbean: At World's End,6.9,4500,4763,2007,May,Saturday
2,43599,245000000,107,880674609,Spectre,6.3,4466,4764,2015,Oct,Monday
3,43600,250000000,112,1084939099,The Dark Knight Rises,7.6,9106,4765,2012,Jul,Monday
4,43602,258000000,115,890871626,Spider-Man 3,5.9,3576,4767,2007,May,Tuesday


`index_col=0` explicitly states to treat the first column as the index.

The default value is `index_col=None`

In [ ]:
# Display the shape of the DataFrame (number of rows, number of columns)
movies.shape

(1465, 11)

The `movies` dataframe contains 1465 rows and 11 columns.

In [ ]:
directors = pd.read_csv('directors.csv', index_col=0) # Use the first column as the index
# Display the first 5 rows of the DataFrame to get a quick overview
print("\n--- Directors DataFrame (First 5 Rows) ---")
display(directors.head())

# Get a concise summary of the DataFrame, including data types and non-null values
print("\n--- Directors DataFrame Info ---")
directors.info()

# Generate descriptive statistics of the DataFrame's numerical columns
print("\n--- Directors DataFrame Descriptive Statistics ---")
display(directors.describe())

# Check for missing values in each column
print("\n--- Directors DataFrame Missing Values ---")
display(directors.isnull().sum())


--- Directors DataFrame (First 5 Rows) ---


,director_name,id,gender
0,James Cameron,4762,Male
1,Gore Verbinski,4763,Male
2,Sam Mendes,4764,Male
3,Christopher Nolan,4765,Male
4,Andrew Stanton,4766,Male



--- Directors DataFrame Info ---
<class 'pandas.core.frame.DataFrame'>
Index: 2349 entries, 0 to 2348
Data columns (total 3 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   director_name  2349 non-null   object
 1   id             2349 non-null   int64 
 2   gender         1724 non-null   object
dtypes: int64(1), object(2)
memory usage: 73.4+ KB

--- Directors DataFrame Descriptive Statistics ---


,id
count,2349.000000
mean,5936.000000
std,678.242213
min,4762.000000
25%,5349.000000
50%,5936.000000
75%,6523.000000
max,7110.000000



--- Directors DataFrame Missing Values ---


,0
director_name,0
id,0
gender,625


In [ ]:
# Display the shape of the directors DataFrame (number of rows, number of columns)
directors.shape

(2349, 3)

The `directors` dataframe contains 2349 rows and 3 columns.

---

### Merging Datasets: Connecting Movies with Directors

To gain a deeper understanding of our movie data, we often need to combine information from different sources. Our `movies` dataset contains `director_id`, but to know the director's name and gender, we need to link it with the `directors` dataset.

**Question:** How can we link the `movies` and `directors` datasets?

We need a common column (a **key**) that exists in both DataFrames. Observing both datasets, we can see that:
- The `movies` DataFrame has a `director_id` column.
- The `directors` DataFrame has an `id` column, which represents the unique identifier for each director.

It's evident that `director_id` in `movies` corresponds to `id` in `directors`. Therefore, we will use these two columns as our **keys** to merge the datasets.

Before merging, let's explore the number of unique directors in each dataset. This will help us understand the relationship between the two DataFrames.

**How can we find the number of unique directors in the `movies` dataset?**

In [ ]:
# Count the number of unique director IDs in the movies DataFrame
movies['director_id'].unique()

array([4762, 4763, 4764, 4765, 4767, 4771, 4772, 4773, 4774, 4775, 4776,
       4777, 4779, 4780, 4782, 4785, 4786, 4788, 4791, 4793, 4794, 4797,
       4798, 4799, 4802, 4803, 4804, 4805, 4808, 4809, 4810, 4811, 4812,
       4813, 4814, 4829, 4831, 4832, 4833, 4838, 4839, 4842, 4845, 4846,
       4847, 4849, 4854, 4857, 4858, 4859, 4863, 4865, 4866, 4867, 4868,
       4872, 4873, 4874, 4876, 4880, 4881, 4885, 4886, 4888, 4889, 4890,
       4892, 4893, 4894, 4900, 4904, 4909, 4913, 4915, 4917, 4918, 4921,
       4923, 4924, 4925, 4926, 4927, 4930, 4935, 4936, 4945, 4947, 4948,
       4949, 4955, 4958, 4960, 4964, 4965, 4967, 4974, 4978, 4989, 4991,
       4993, 4995, 4998, 5002, 5005, 5008, 5009, 5014, 5017, 5024, 5031,
       5033, 5034, 5035, 5037, 5039, 5047, 5053, 5056, 5059, 5061, 5067,
       5082, 5087, 5088, 5093, 5096, 5097, 5101, 5105, 5111, 5113, 5115,
       5116, 5123, 5125, 5130, 5132, 5135, 5140, 5147, 5148, 5152, 5157,
       5162, 5179, 5182, 5191, 5193, 5195, 5196, 52

In [ ]:
directors["id"].unique()

array([4762, 4763, 4764, ..., 7108, 7109, 7110])

In [ ]:
movies_dir_merged = movies.merge(directors, how="left", left_on="director_id", right_on="id")

In [ ]:
movies_dir_merged.head()

,Unnamed: 0,id_x,budget,popularity,revenue,title,vote_average,vote_count,director_id,year,month,day,director_name,id_y,gender
0,0,43597,237000000,150,2787965087,Avatar,7.2,11800,4762,2009,Dec,Thursday,James Cameron,4762,Male
1,1,43598,300000000,139,961000000,Pirates of the Caribbean: At World's End,6.9,4500,4763,2007,May,Saturday,Gore Verbinski,4763,Male
2,2,43599,245000000,107,880674609,Spectre,6.3,4466,4764,2015,Oct,Monday,Sam Mendes,4764,Male
3,3,43600,250000000,112,1084939099,The Dark Knight Rises,7.6,9106,4765,2012,Jul,Monday,Christopher Nolan,4765,Male
4,5,43602,258000000,115,890871626,Spider-Man 3,5.9,3576,4767,2007,May,Tuesday,Sam Raimi,4767,Male


In [ ]:
movies_dir_merged.drop(["id_y"], axis=1, inplace=True)

In [ ]:
movies_dir_merged.head()

,Unnamed: 0,id_x,budget,popularity,revenue,title,vote_average,vote_count,director_id,year,month,day,director_name,gender
0,0,43597,237000000,150,2787965087,Avatar,7.2,11800,4762,2009,Dec,Thursday,James Cameron,Male
1,1,43598,300000000,139,961000000,Pirates of the Caribbean: At World's End,6.9,4500,4763,2007,May,Saturday,Gore Verbinski,Male
2,2,43599,245000000,107,880674609,Spectre,6.3,4466,4764,2015,Oct,Monday,Sam Mendes,Male
3,3,43600,250000000,112,1084939099,The Dark Knight Rises,7.6,9106,4765,2012,Jul,Monday,Christopher Nolan,Male
4,5,43602,258000000,115,890871626,Spider-Man 3,5.9,3576,4767,2007,May,Tuesday,Sam Raimi,Male


In [ ]:
columns = {"id_x" : "movies_id"}
movies_dir_merged.rename(columns=columns, inplace=True)

In [ ]:
movies_dir_merged.head()

,Unnamed: 0,movies_id,budget,popularity,revenue,title,vote_average,vote_count,director_id,year,month,day,director_name,gender
0,0,43597,237000000,150,2787965087,Avatar,7.2,11800,4762,2009,Dec,Thursday,James Cameron,Male
1,1,43598,300000000,139,961000000,Pirates of the Caribbean: At World's End,6.9,4500,4763,2007,May,Saturday,Gore Verbinski,Male
2,2,43599,245000000,107,880674609,Spectre,6.3,4466,4764,2015,Oct,Monday,Sam Mendes,Male
3,3,43600,250000000,112,1084939099,The Dark Knight Rises,7.6,9106,4765,2012,Jul,Monday,Christopher Nolan,Male
4,5,43602,258000000,115,890871626,Spider-Man 3,5.9,3576,4767,2007,May,Tuesday,Sam Raimi,Male


The `.nunique()` method is a powerful Series function in Pandas that returns the number of unique values in a column. It's very useful for understanding the diversity or cardinality of a particular categorical feature. In this case, `movies['director_id'].nunique()` tells us how many distinct directors are represented in our `movies` dataset.

In [ ]:
# Count the number of unique director IDs in the directors DataFrame
directors['id'].nunique()

2349

### Comparing Director Counts

Let's summarize our findings:
- The `movies` dataset contains 1465 movie entries, but only 199 unique `director_id`s.
- The `directors` dataset lists 2349 unique `id`s (which is equal to its number of rows).

**What can we infer from this observation?**
This indicates that not all directors listed in the `directors` dataset have movies present in our `movies` dataset. The set of directors in `movies` is a subset of the directors in the `directors` dataset.

**How can we confirm that all `director_id` values in `movies` are indeed present in the `id` column of `directors`?**

In [ ]:
# Check if each director_id in 'movies' is present in the 'id' column of 'directors'
condition = movies['director_id'].isin(directors['id'])

,Unnamed: 0,id,budget,popularity,revenue,title,vote_average,vote_count,director_id,year,month,day
0,0,43597,237000000,150,2787965087,Avatar,7.2,11800,4762,2009,Dec,Thursday
1,1,43598,300000000,139,961000000,Pirates of the Caribbean: At World's End,6.9,4500,4763,2007,May,Saturday
2,2,43599,245000000,107,880674609,Spectre,6.3,4466,4764,2015,Oct,Monday
3,3,43600,250000000,112,1084939099,The Dark Knight Rises,7.6,9106,4765,2012,Jul,Monday
4,5,43602,258000000,115,890871626,Spider-Man 3,5.9,3576,4767,2007,May,Tuesday
...,...,...,...,...,...,...,...,...,...,...,...,...
1460,4736,48363,0,3,321952,The Last Waltz,7.9,64,4809,1978,May,Monday
1461,4743,48370,27000,19,3151130,Clerks,7.4,755,5369,1994,Sep,Tuesday
1462,4748,48375,0,7,0,Rampage,6.0,131,5148,2009,Aug,Friday
1463,4749,48376,0,3,0,Slacker,6.4,77,5535,1990,Jul,Friday


### Understanding `isin()` for Membership Testing

The `.isin()` method is used to check if each element in a Series is contained in a provided list or Series of values. It's a very efficient way to perform membership testing across an entire column.

**How is `isin()` different from Python's standard `in` operator?**
- Python's `in` operator typically works for checking membership of **a single element** within a collection.
- Pandas' `.isin()` method performs this check for **all values** in a Series simultaneously, making it highly efficient for DataFrame operations.

As you can see, the output of `movies['director_id'].isin(directors['id'])` is a **boolean mask**. This is a Series of `True`/`False` values, where `True` indicates that the `director_id` at that position is found in the `directors['id']` column, and `False` if it's not.

**How can we easily check if *any* `False` values exist in this boolean mask?** (i.e., if there's any `director_id` in `movies` that *doesn't* have a match in `directors`)

In [ ]:
directors['id']

,id
0,4762
1,4763
2,4764
3,4765
4,4766
...,...
2344,7106
2345,7107
2346,7108
2347,7109


In [ ]:
movies['director_id'].isin(directors['id'])

,director_id
0,True
1,True
2,True
3,True
4,True
...,...
1460,True
1461,True
1462,True
1463,True


In [ ]:
# Check if ALL director_id values inmovies['director_id'].isin(directors['id']) 'movies' are present in the 'id' column of 'directors'
# np.all() returns True if all elements in the boolean Series are True.
np.all(movies['director_id'].isin(directors['id']))

np.True_

### Deciding on the Merge Strategy: Left Join

Now that we've confirmed the relationship between our two datasets, let's merge them. The choice of merge type is crucial and depends on what we want to achieve:

- **Do we need to keep all the rows from the `movies` DataFrame?** Yes, because we want to analyze all movies.
- **Do we need to keep all the rows from the `directors` DataFrame?** No. We only need the director information for those directors who actually directed a movie in our `movies` dataset.

Based on these requirements, a **`LEFT` Join** is the most appropriate merge type. A left join keeps all rows from the 'left' DataFrame (our `movies` DataFrame) and matches them with rows from the 'right' DataFrame (our `directors` DataFrame) based on the specified keys (`director_id` and `id`). If a match isn't found in the right DataFrame, the columns from the right DataFrame will have `NaN` (Not a Number) values for that row.

In [ ]:
# Perform a left merge of 'movies' and 'directors' DataFrames
# 'left_on' specifies the column from the left DataFrame ('movies')
# 'right_on' specifies the column from the right DataFrame ('directors')
# 'how='left'' ensures all rows from 'movies' are kept.
data = movies.merge(directors, how='left', left_on='director_id',right_on='id')
# Display the merged DataFrame
data

,id_x,budget,popularity,revenue,title,vote_average,vote_count,director_id,year,month,day,director_name,id_y,gender
0,43597,237000000,150,2787965087,Avatar,7.2,11800,4762,2009,Dec,Thursday,James Cameron,4762,Male
1,43598,300000000,139,961000000,Pirates of the Caribbean: At World's End,6.9,4500,4763,2007,May,Saturday,Gore Verbinski,4763,Male
2,43599,245000000,107,880674609,Spectre,6.3,4466,4764,2015,Oct,Monday,Sam Mendes,4764,Male
3,43600,250000000,112,1084939099,The Dark Knight Rises,7.6,9106,4765,2012,Jul,Monday,Christopher Nolan,4765,Male
4,43602,258000000,115,890871626,Spider-Man 3,5.9,3576,4767,2007,May,Tuesday,Sam Raimi,4767,Male
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1460,48363,0,3,321952,The Last Waltz,7.9,64,4809,1978,May,Monday,Martin Scorsese,4809,Male
1461,48370,27000,19,3151130,Clerks,7.4,755,5369,1994,Sep,Tuesday,Kevin Smith,5369,Male
1462,48375,0,7,0,Rampage,6.0,131,5148,2009,Aug,Friday,Uwe Boll,5148,Male
1463,48376,0,3,0,Slacker,6.4,77,5535,1990,Jul,Friday,Richard Linklater,5535,Male


### Understanding and Cleaning Merged Columns (`id_x`, `id_y`)

Upon merging, you might have noticed two new columns: `id_x` and `id_y`.

**What are these newly created columns?**
When two DataFrames are merged and both contain columns with the same name (in this case, `id`), Pandas automatically renames them to prevent conflicts. By default, it appends `_x` to the column from the left DataFrame (`movies`) and `_y` to the column from the right DataFrame (`directors`).

- `id_x` represents the unique ID for each movie from the original `movies` DataFrame.
- `id_y` represents the `id` of the director from the `directors` DataFrame that matched the `director_id` in the `movies` DataFrame.

**Are any of these columns redundant and can be safely removed?**
- The `id_y` column essentially duplicates the `director_id` column after the merge, as it contains the same director ID used for the join. Once the merge is complete, `id_y` doesn't add new, unique information beyond what `director_id` already provides.
- While `director_id` served its purpose for the merge, we might not need it for subsequent analysis if we're now working with `director_name`.

Thus, we can safely drop both `director_id` (since we have `director_name` linked) and `id_y` to keep our DataFrame clean and avoid redundant information.

In [ ]:
# Drop the redundant 'director_id' and 'id_y' columns from the merged DataFrame
# 'axis=1' specifies that columns should be dropped
# 'inplace=True' modifies the DataFrame directly without needing re-assignment
data.drop(['director_id','id_y'],axis=1,inplace=True)
# Display the first few rows of the cleaned DataFrame
data.head()

,id_x,budget,popularity,revenue,title,vote_average,vote_count,year,month,day,director_name,gender
0,43597,237000000,150,2787965087,Avatar,7.2,11800,2009,Dec,Thursday,James Cameron,Male
1,43598,300000000,139,961000000,Pirates of the Caribbean: At World's End,6.9,4500,2007,May,Saturday,Gore Verbinski,Male
2,43599,245000000,107,880674609,Spectre,6.3,4466,2015,Oct,Monday,Sam Mendes,Male
3,43600,250000000,112,1084939099,The Dark Knight Rises,7.6,9106,2012,Jul,Monday,Christopher Nolan,Male
4,43602,258000000,115,890871626,Spider-Man 3,5.9,3576,2007,May,Tuesday,Sam Raimi,Male


In [ ]:
data.rename(columns={"id_x": "movies_id"}, inplace=True)

In [ ]:
data.head()

,movies_id,budget,popularity,revenue,title,vote_average,vote_count,year,month,day,director_name,gender
0,43597,237000000,150,2787965087,Avatar,7.2,11800,2009,Dec,Thursday,James Cameron,Male
1,43598,300000000,139,961000000,Pirates of the Caribbean: At World's End,6.9,4500,2007,May,Saturday,Gore Verbinski,Male
2,43599,245000000,107,880674609,Spectre,6.3,4466,2015,Oct,Monday,Sam Mendes,Male
3,43600,250000000,112,1084939099,The Dark Knight Rises,7.6,9106,2012,Jul,Monday,Christopher Nolan,Male
4,43602,258000000,115,890871626,Spider-Man 3,5.9,3576,2007,May,Tuesday,Sam Raimi,Male


In [ ]:
data.sort_values(by="popularity", ascending=False)[["title", "popularity"]].head().reset_index()

,index,title,popularity
0,58,Interstellar,724
1,78,Mad Max: Fury Road,434
2,119,Pirates of the Caribbean: The Curse of the Bla...,271
3,120,The Hunger Games: Mockingjay - Part 1,206
4,45,The Dark Knight,187


In [ ]:
# data["revenue"] - data["budget"]

## Post-Merge Exploration and Your Turn!

Now that our `movies` and `directors` datasets are successfully merged into a single DataFrame named `data`, we have a much richer dataset to work with. This combined view opens up many possibilities for data exploration and insight extraction.

From here, you have the opportunity to delve into various aspects of the data, such as:

-   **Financial Performance:** Calculate profit/loss, convert revenue values into Millions of USD for better readability.
-   **Top Performers:** Identify the Top 5 most popular movies, or the directors with the highest average revenue.
-   **Demographic Analysis:** Investigate if there are differences in movie performance based on director gender.

**Your Task:** Explore the merged data on your own! Try to answer some of these questions using the Pandas methods you've learned so far. This hands-on exploration will solidify your understanding of data manipulation.


### Post-read

- [IMDB data exploration](https://colab.research.google.com/drive/1yrfHSQYUMxxLKGUG-gCPf-R232BuimiR?usp=sharing)

From here, we have the opportunity to delve into various aspects of the data, such as:

- Converting the revenue values into Millions of USD.
- Identifying the Top 5 most popular movies.

... and so on.

This task is for you to explore the data on your own.


---

### Quiz-1

# Question
How to filter those records where movies are released either in the year 2015, 2016 or 2012 from the `movies` dataset?

# Choices
- [ ] `df['year'].isin([2015, 2016, 2012])`
- [ ] `df['year'].in([2015, 2016, 2012])`
- [ ] `df['year']==([2015, 2016, 2012])`

In [ ]:
data["year"]

,year
0,2009
1,2007
2,2015
3,2012
4,2007
...,...
1460,1978
1461,1994
1462,2009
1463,1990


---

## Mastering the `.apply()` Method: Custom Transformations

The `.apply()` method is an incredibly versatile function in Pandas that allows you to apply a custom function along either axis (rows or columns) of a DataFrame or Series. This is particularly useful when you need to perform operations that aren't built-in Pandas functions.

**Scenario:** Let's say we want to convert the categorical values in the `gender` column (e.g., 'Male', 'Female') into a numerical format. This process is called **encoding**, and it's a common step in data preparation for machine learning models.

**Goal:**
- Encode 'Male' as `0`
- Encode 'Female' as `1`

**How can we encode the values in the `gender` column using a custom function?**
First, let's write a simple Python function that performs this encoding for a single value.

In [ ]:
data.head()

,movies_id,budget,popularity,revenue,title,vote_average,vote_count,year,month,day,director_name,gender
0,43597,237000000,150,2787965087,Avatar,7.2,11800,2009,Dec,Thursday,James Cameron,Male
1,43598,300000000,139,961000000,Pirates of the Caribbean: At World's End,6.9,4500,2007,May,Saturday,Gore Verbinski,Male
2,43599,245000000,107,880674609,Spectre,6.3,4466,2015,Oct,Monday,Sam Mendes,Male
3,43600,250000000,112,1084939099,The Dark Knight Rises,7.6,9106,2012,Jul,Monday,Christopher Nolan,Male
4,43602,258000000,115,890871626,Spider-Man 3,5.9,3576,2007,May,Tuesday,Sam Raimi,Male


In [ ]:
# Define a function to encode gender values
def encode(data):
  if data == "Male":
    return 0  # Encode 'Male' as 0
  elif data == "Female":
    return 1  # Encode 'Female' as 1
  else:
    return "Not Found"

Now that we have our `encode` function, **how can we apply this function to *every* value in the `gender` column of our `data` DataFrame?** The `.apply()` method is perfect for this.

In [ ]:
# Apply the 'encode' function to the 'gender' column
# This transforms 'Male'/'Female' strings into 0/1 integers
data['gender'] = data['gender'].apply(encode)
# Display the updated DataFrame to see the encoded 'gender' column
data

,movies_id,budget,popularity,revenue,title,vote_average,vote_count,year,month,day,director_name,gender
0,43597,237000000,150,2787965087,Avatar,7.2,11800,2009,Dec,Thursday,James Cameron,0
1,43598,300000000,139,961000000,Pirates of the Caribbean: At World's End,6.9,4500,2007,May,Saturday,Gore Verbinski,0
2,43599,245000000,107,880674609,Spectre,6.3,4466,2015,Oct,Monday,Sam Mendes,0
3,43600,250000000,112,1084939099,The Dark Knight Rises,7.6,9106,2012,Jul,Monday,Christopher Nolan,0
4,43602,258000000,115,890871626,Spider-Man 3,5.9,3576,2007,May,Tuesday,Sam Raimi,0
...,...,...,...,...,...,...,...,...,...,...,...,...
1460,48363,0,3,321952,The Last Waltz,7.9,64,1978,May,Monday,Martin Scorsese,0
1461,48370,27000,19,3151130,Clerks,7.4,755,1994,Sep,Tuesday,Kevin Smith,0
1462,48375,0,7,0,Rampage,6.0,131,2009,Aug,Friday,Uwe Boll,0
1463,48376,0,3,0,Slacker,6.4,77,1990,Jul,Friday,Richard Linklater,0


In [ ]:
"""
Monday = 0
Tuesday = 1
Wednesday = 2
Thursday = 3
Friday = 4
Saturday = 5
Sunday = 6
"""

# 1. Create a function
def encode_days(data):
  if data == "Monday":
    return 0
  elif data == "Tuesday":
    return 1
  elif data == "Wednesday":
    return 2
  ...


data["day"].apply(encode_days)

### Applying Functions on Multiple Columns: Understanding `axis`

Notice how using `.apply()` with our custom `encode` function is conceptually similar to `Vectorization` in NumPy, where an operation is applied efficiently across an entire array or Series.

**Question:** What if we want to apply a function that involves *multiple* columns? For example, how can we find the sum of `revenue` and `budget` for each movie?

If we try to apply a sum directly to multiple columns, as shown in the previous cell, we might get an unexpected result. We want the sum *per movie* (i.e., per row), but the default behavior sums *down* the columns.

In [ ]:
# def profit()

In [ ]:
data[["revenue", "budget"]]

,revenue,budget
0,2787965087,237000000
1,961000000,300000000
2,880674609,245000000
3,1084939099,250000000
4,890871626,258000000
...,...,...
1460,321952,0
1461,3151130,27000
1462,0,0
1463,0,0


In [ ]:
# Attempt to apply the sum function to 'revenue' and 'budget' columns
# By default, apply() with axis=0 sums down each column separately
data[['revenue', 'budget']].apply(np.sum, axis=0)

,0
revenue,209866997305
budget,70353617179


### The Importance of the `axis` Parameter

As you saw, applying `np.sum` to `data[['revenue', 'budget']]` without specifying the `axis` parameter resulted in the sum of all `revenue` values and the sum of all `budget` values separately (summing along `axis=0`, which is the default for columns).

To achieve our goal of summing `revenue` and `budget` *for each movie* (i.e., *per row*), we need to explicitly tell Pandas to apply the function along `axis=1`.

**Understanding `axis` in `.apply()`:**
- **`axis=0`**: Applies the function **column-wise**. This is the default. The function receives each column as a Series.
- **`axis=1`**: Applies the function **row-wise**. The function receives each row as a Series.

By setting `axis=1`, our `np.sum` function now receives each row (containing `revenue` and `budget` for a single movie) and correctly calculates their sum.

**Your Turn:** Similarly, how can you calculate the `profit` for each movie (where `profit = revenue - budget`) and add it as a new column to our DataFrame? Think about how to define a function that takes a row and returns the profit.

In [ ]:
# Apply the sum function across rows (axis=1) for 'revenue' and 'budget'
# This calculates the sum of revenue and budget for each movie
data[['revenue', 'budget']].apply(np.sum, axis=1)

,0
0,3024965087
1,1261000000
2,1125674609
3,1334939099
4,1148871626
...,...
1460,321952
1461,3178130
1462,0
1463,0


By setting the `axis=1`, every row of `revenue` was added to same row of `budget`.

**What does this `axis` mean in apply?**
- `axis=0` $\rightarrow$ It will apply to **each column**
- `axis=1` $\rightarrow$ It will apply to **each row**
  
Note that **by default, axis=0**.

**Similarly, how can I find the `profit` per movie (revenue-budget)?**

In [ ]:
x = data.loc[2]

In [ ]:
x['revenue'] - x['budget']

np.int64(635674609)

In [ ]:
# Define a function to calculate profit for a given row (x)
def prof(x):
  # Profit is calculated as revenue minus budget
  return x['revenue']-x['budget']

# Apply the 'prof' function to each row (axis=1) of the 'revenue' and 'budget' columns
# Store the result in a new 'profit' column
data['profit'] = data[['revenue', 'budget']].apply(prof, axis=1)
# Display the updated DataFrame including the new 'profit' column
data

,movies_id,budget,popularity,revenue,title,vote_average,vote_count,year,month,day,director_name,gender,profit
0,43597,237000000,150,2787965087,Avatar,7.2,11800,2009,Dec,Thursday,James Cameron,0,2550965087
1,43598,300000000,139,961000000,Pirates of the Caribbean: At World's End,6.9,4500,2007,May,Saturday,Gore Verbinski,0,661000000
2,43599,245000000,107,880674609,Spectre,6.3,4466,2015,Oct,Monday,Sam Mendes,0,635674609
3,43600,250000000,112,1084939099,The Dark Knight Rises,7.6,9106,2012,Jul,Monday,Christopher Nolan,0,834939099
4,43602,258000000,115,890871626,Spider-Man 3,5.9,3576,2007,May,Tuesday,Sam Raimi,0,632871626
...,...,...,...,...,...,...,...,...,...,...,...,...,...
1460,48363,0,3,321952,The Last Waltz,7.9,64,1978,May,Monday,Martin Scorsese,0,321952
1461,48370,27000,19,3151130,Clerks,7.4,755,1994,Sep,Tuesday,Kevin Smith,0,3124130
1462,48375,0,7,0,Rampage,6.0,131,2009,Aug,Friday,Uwe Boll,0,0
1463,48376,0,3,0,Slacker,6.4,77,1990,Jul,Friday,Richard Linklater,0,0


### Exercise 2: Custom Transformations with `apply()`

Now that you've seen how to use `apply()` for encoding and calculating profit, let's try another transformation.

**Task:**
1.  Create a new column named `is_successful` in the `data` DataFrame.
2.  This column should be a boolean (`True`/`False`) indicating whether a movie made a profit (i.e., `profit > 0`).
3.  Display the `title`, `revenue`, `budget`, `profit`, and `is_successful` columns for the first 10 rows.

*Hint: You can use a lambda function directly within `apply()` for simple conditions or define a separate function if the logic is more complex.*

In [ ]:
# 1. Create a function
def check_if_successful(x):
  if x["profit"] > 0:
    return True
  return False

data["is_successful_method1"] = data.apply(check_if_successful, axis=1)

data["is_successful_method2"] = data.apply(lambda x: x["profit"] > 0, axis=1)

In [ ]:
data.head()

,movies_id,budget,popularity,revenue,title,vote_average,vote_count,year,month,day,director_name,gender,profit,is_successful_method1,is_successful_method2
0,43597,237000000,150,2787965087,Avatar,7.2,11800,2009,Dec,Thursday,James Cameron,0,2550965087,True,True
1,43598,300000000,139,961000000,Pirates of the Caribbean: At World's End,6.9,4500,2007,May,Saturday,Gore Verbinski,0,661000000,True,True
2,43599,245000000,107,880674609,Spectre,6.3,4466,2015,Oct,Monday,Sam Mendes,0,635674609,True,True
3,43600,250000000,112,1084939099,The Dark Knight Rises,7.6,9106,2012,Jul,Monday,Christopher Nolan,0,834939099,True,True
4,43602,258000000,115,890871626,Spider-Man 3,5.9,3576,2007,May,Tuesday,Sam Raimi,0,632871626,True,True


In [ ]:
# 1. Create the 'is_successful' column: True if profit > 0, False otherwise
data['is_successful'] = data['profit'].apply(lambda x: x > 0)

# 2. Display the requested columns for the first 10 rows
print("\n--- First 10 rows with 'is_successful' status ---")
display(data[['title', 'revenue', 'budget', 'profit', 'is_successful']].head(10))


--- First 10 rows with 'is_successful' status ---


,title,revenue,budget,profit,is_successful
0,Avatar,2787965087,237000000,2550965087,True
1,Pirates of the Caribbean: At World's End,961000000,300000000,661000000,True
2,Spectre,880674609,245000000,635674609,True
3,The Dark Knight Rises,1084939099,250000000,834939099,True
4,Spider-Man 3,890871626,258000000,632871626,True
5,Batman v Superman: Dawn of Justice,873260194,250000000,623260194,True
6,Superman Returns,391081192,270000000,121081192,True
7,Quantum of Solace,586090727,200000000,386090727,True
8,Pirates of the Caribbean: Dead Man's Chest,1065659812,200000000,865659812,True
9,The Lone Ranger,89289910,255000000,-165710090,False


## The Power of `groupby()`: Split, Apply, Combine

The `groupby()` method is one of the most powerful and frequently used functions in Pandas. It allows you to group rows together based on common values in one or more columns, and then perform operations on these groups independently.

The process of `groupby()` can be conceptualized as three distinct steps:

1.  **Split**: Pandas divides the DataFrame into smaller

## Group-based Operations: Aggregation

After splitting the DataFrame into groups using `groupby()`, the next step is often to **Apply** an aggregation function to each group. Aggregation functions condense the data in each group into a single summary statistic.

Common aggregate functions include:
-   `mean()`: Calculates the average of values in each group.
-   `sum()`: Calculates the total sum of values in each group.
-   `min()`: Finds the minimum value in each group.
-   `max()`: Finds the maximum value in each group.
-   `count()`: Counts the number of non-null items in each group.

**Let's begin by grouping our `data` DataFrame by `director_name`.** This will allow us to analyze metrics for each director.

### Group based Aggregation

We use different aggregate functions like `mean`, `sum`, `min`, `max`, `count` etc. on columns while grouping.

Let's group our data director-wise.

In [ ]:
data.head()

,movies_id,budget,popularity,revenue,title,vote_average,vote_count,year,month,day,director_name,gender,profit,is_successful_method1,is_successful_method2
0,43597,237000000,150,2787965087,Avatar,7.2,11800,2009,Dec,Thursday,James Cameron,0,2550965087,True,True
1,43598,300000000,139,961000000,Pirates of the Caribbean: At World's End,6.9,4500,2007,May,Saturday,Gore Verbinski,0,661000000,True,True
2,43599,245000000,107,880674609,Spectre,6.3,4466,2015,Oct,Monday,Sam Mendes,0,635674609,True,True
3,43600,250000000,112,1084939099,The Dark Knight Rises,7.6,9106,2012,Jul,Monday,Christopher Nolan,0,834939099,True,True
4,43602,258000000,115,890871626,Spider-Man 3,5.9,3576,2007,May,Tuesday,Sam Raimi,0,632871626,True,True


In [ ]:
data.groupby('director_name')

In [ ]:
# Group the DataFrame by 'director_name'
# This creates a DataFrameGroupBy object, not a DataFrame itself
data.groupby('director_name')["profit"].mean() # director -> 1 -> 1 Million USD ,  0 USD; 2 -> 500K, 500k, 500k, 500K

,profit
director_name,
Adam McKay,8.626274e+07
Adam Shankman,6.082035e+07
Alejandro González Iñárritu,1.129966e+08
Alex Proyas,8.451249e+07
Alexander Payne,6.644699e+07
...,...
Wes Craven,5.309943e+07
Wolfgang Petersen,1.405737e+08
Woody Allen,2.271702e+07


In [ ]:
data.groupby('director_name')["movies_id"].count().sort_values(ascending=False)

,movies_id
director_name,
Steven Spielberg,26
Martin Scorsese,19
Clint Eastwood,19
Woody Allen,18
Robert Rodriguez,16
...,...
Stephen Daldry,5
Tom Tykwer,5
Tim Hill,5


### Understanding the `DataFrameGroupBy` Object

When you call `data.groupby('director_name')`, it doesn't immediately return a new DataFrame. Instead, it returns a `DataFrameGroupBy` object. This object is an intermediate step that holds the grouping information but hasn't yet applied any operations.

- It's a special object that knows *how* to group your data.
- It's **not a DataFrame** itself; you need to apply an aggregation or transformation function to it to get a DataFrame or Series back.

**Question:** How can we find out the total number of distinct groups (i.e., the number of unique directors) our data has been divided into?

In [ ]:
# Get the number of distinct groups (unique directors) after grouping
data.groupby('director_name').ngroups # Number of Groups

199

### Inspecting Group Contents: `groups` and `get_group()`

After understanding the number of groups, we can also inspect the actual composition of these groups. The `.groups` attribute of a `DataFrameGroupBy` object returns a dictionary where keys are the group names (e.g., director names) and values are the indices of the rows belonging to that group.

**Question:** Based on this grouping, how can we see which specific rows (movies) belong to a particular director's group? For instance, how would we view all movies directed by 'Alexander Payne'?

In [ ]:
# View the 'groups' attribute of the groupby object
# This shows a dictionary where keys are group names and values are row indices
data.groupby('director_name').groups

{'Adam McKay': [176, 323, 366, 505, 839, 916], 'Adam Shankman': [265, 300, 350, 404, 458, 843, 999, 1231], 'Alejandro González Iñárritu': [106, 749, 1015, 1034, 1077, 1405], 'Alex Proyas': [95, 159, 514, 671, 873], 'Alexander Payne': [793, 1006, 1101, 1211, 1281], 'Andrew Adamson': [11, 43, 328, 501, 947], 'Andrew Niccol': [533, 603, 701, 722, 1439], 'Andrzej Bartkowiak': [349, 549, 754, 911, 924], 'Andy Fickman': [517, 681, 909, 926, 973, 1023], 'Andy Tennant': [314, 320, 464, 593, 676, 885], 'Ang Lee': [99, 134, 748, 840, 1089, 1110, 1132, 1184], 'Anne Fletcher': [610, 650, 736, 789, 1206], 'Antoine Fuqua': [310, 338, 424, 467, 576, 808, 818, 1105], 'Atom Egoyan': [946, 1128, 1164, 1194, 1347, 1416], 'Barry Levinson': [313, 319, 471, 594, 878, 898, 1013, 1037, 1082, 1143, 1185, 1345, 1378], 'Barry Sonnenfeld': [13, 48, 90, 205, 591, 778, 783], 'Ben Stiller': [209, 212, 547, 562, 850], 'Bill Condon': [102, 307, 902, 1233, 1381], 'Bobby Farrelly': [352, 356, 481, 498, 624, 630, 654, 80

**What if we want to extract data of a particular group from this list?**

In [ ]:
# Extract all rows (movies) belonging to the 'Alexander Payne' group
data.groupby('director_name').get_group('Steven Spielberg')

,movies_id,budget,popularity,revenue,title,vote_average,vote_count,year,month,day,director_name,gender,profit,is_successful_method1,is_successful_method2
37,43650,185000000,75,786636033,Indiana Jones and the Kingdom of the Crystal S...,5.7,2495,2008,May,Wednesday,Steven Spielberg,0,601636033,True,True
105,43772,140000000,44,183345589,The BFG,6.0,1000,2016,Jun,Wednesday,Steven Spielberg,0,43345589,True,True
110,43782,132000000,48,591739379,War of the Worlds,6.2,2322,2005,Jun,Tuesday,Steven Spielberg,0,459739379,True,True
114,43787,130000000,89,371940071,The Adventures of Tintin,6.7,2061,2011,Oct,Tuesday,Steven Spielberg,0,241940071,True,True
166,43872,102000000,65,358372926,Minority Report,7.1,2608,2002,Jun,Thursday,Steven Spielberg,0,256372926,True,True
219,43960,100000000,34,235926552,A.I. Artificial Intelligence,6.8,1974,2001,Jun,Friday,Steven Spielberg,0,135926552,True,True
296,44105,73000000,2,229074524,The Lost World: Jurassic Park,6.2,2487,1997,May,Friday,Steven Spielberg,0,156074524,True,True
304,44118,60000000,57,219417255,The Terminal,7.0,1910,2004,Jun,Thursday,Steven Spielberg,0,159417255,True,True
309,44125,70000000,29,130358911,Munich,6.9,696,2005,Dec,Thursday,Steven Spielberg,0,60358911,True,True
333,44169,70000000,33,300854823,Hook,6.6,1532,1991,Dec,Wednesday,Steven Spielberg,0,230854823,True,True


### Counting Movies per Director

Now, let's use aggregation to answer a practical question: **How can we find the total count of movies directed by each director?**

To do this, we can select a column (like `title`) after grouping by `director_name` and then apply the `.count()` aggregation function. This will count the non-null movie titles for each director.

In [ ]:
data.groupby('director_name')['profit'].mean()

,profit
director_name,
Adam McKay,8.626274e+07
Adam Shankman,6.082035e+07
Alejandro González Iñárritu,1.129966e+08
Alex Proyas,8.451249e+07
Alexander Payne,6.644699e+07
...,...
Wes Craven,5.309943e+07
Wolfgang Petersen,1.405737e+08
Woody Allen,2.271702e+07


In [ ]:
# Group by 'director_name' and count the number of 'title' entries for each director
# This effectively counts the number of movies directed by each director
data.groupby('director_name')['title'].count()

,title
director_name,
Adam McKay,6
Adam Shankman,8
Alejandro González Iñárritu,6
Alex Proyas,5
Alexander Payne,5
...,...
Wes Craven,10
Wolfgang Petersen,7
Woody Allen,18


### Multiple Aggregations with `.aggregate()` (or `.agg()`)

Sometimes, you need to calculate several different summary statistics for the same grouped data. Pandas' `.aggregate()` (or its alias `.agg()`) method allows you to apply multiple aggregation functions at once.

**Scenario:** Let's find the very first year and the latest year a director released a movie. This involves finding the minimum (`min`) and maximum (`max`) values of the `year` column for each `director_name` group.

**Note:** Both `.aggregate()` and `.agg()` perform the same function. `.agg()` is often preferred for its conciseness.

In [ ]:
data.groupby("director_name").agg({"title" : "count", "profit" : "mean", "year": "min"})

,title,profit,year
director_name,,,
Adam McKay,6,8.626274e+07,2004
Adam Shankman,8,6.082035e+07,2001
Alejandro González Iñárritu,6,1.129966e+08,2000
Alex Proyas,5,8.451249e+07,1994
Alexander Payne,5,6.644699e+07,1999
...,...,...,...
Wes Craven,10,5.309943e+07,1984
Wolfgang Petersen,7,1.405737e+08,1981
Woody Allen,18,2.271702e+07,1977


In [ ]:
data.groupby("director_name")["profit"].aggregate(["min", "max", "sum", "mean"])

,min,max,sum,mean
director_name,,,,
Adam McKay,63107642,123649015,517576455,8.626274e+07
Adam Shankman,-15581387,132874442,486562773,6.082035e+07
Alejandro González Iñárritu,18908467,397950503,677979871,1.129966e+08
Alex Proyas,200316,227234916,422562458,8.451249e+07
Alexander Payne,0,157243185,332234956,6.644699e+07
...,...,...,...,...
Wes Craven,-15705099,159046663,530994281,5.309943e+07
Wolfgang Petersen,-6841192,322409852,984016083,1.405737e+08
Woody Allen,-22000000,121119219,408906323,2.271702e+07


In [ ]:
# Group by 'director_name' and calculate the minimum and maximum 'year' for each director
# .aggregate() allows applying multiple functions to a column within groups
data.groupby(['director_name'])["year"].aggregate(['min', 'max'])

,min,max
director_name,,
Adam McKay,2004,2015
Adam Shankman,2001,2012
Alejandro González Iñárritu,2000,2015
Alex Proyas,1994,2016
Alexander Payne,1999,2013
...,...,...
Wes Craven,1984,2011
Wolfgang Petersen,1981,2006
Woody Allen,1977,2013


In [ ]:
data.groupby(['director_name'])["year"].aggregate(['min', 'max']).apply(lambda x : x["max"] - x["min"], axis=1)

,0
director_name,
Adam McKay,11
Adam Shankman,11
Alejandro González Iñárritu,15
Alex Proyas,22
Alexander Payne,14
...,...
Wes Craven,27
Wolfgang Petersen,25
Woody Allen,36


In [ ]:
data.groupby("director_name")["popularity"].agg(["mean", "min", "max", "count"])

,mean,min,max,count
director_name,,,,
Adam McKay,30.333333,12,57,6
Adam Shankman,23.125000,8,35,8
Alejandro González Iñárritu,47.000000,13,100,6
Alex Proyas,53.200000,32,95,5
Alexander Payne,24.800000,13,40,5
...,...,...,...,...
Wes Craven,22.300000,6,45,10
Wolfgang Petersen,35.857143,21,66,7
Woody Allen,17.722222,7,43,18


**Note:** We can also use `.agg` instead of `.aggregate` (both are same)

### Exercise 3: Advanced Grouped Aggregations

Let's apply what you've learned about `groupby()` and aggregation to extract more insights.

**Task:**
1.  Find the `director_name` with the highest total `revenue`.
2.  For each `director_name`, calculate the average `vote_average` and the total number of movies (`title` count).
3.  Present these results in a well-structured DataFrame, sorted by total `revenue` in descending order.

*Hint: You'll likely need to use `.agg()` with a dictionary to specify different aggregations for different columns, and then sort your results.*

In [ ]:
data.groupby("director_name").agg({"title" : "count", "profit" : "mean", "year": "min"})

,title,profit,year
director_name,,,
Adam McKay,6,8.626274e+07,2004
Adam Shankman,8,6.082035e+07,2001
Alejandro González Iñárritu,6,1.129966e+08,2000
Alex Proyas,5,8.451249e+07,1994
Alexander Payne,5,6.644699e+07,1999
...,...,...,...
Wes Craven,10,5.309943e+07,1984
Wolfgang Petersen,7,1.405737e+08,1981
Woody Allen,18,2.271702e+07,1977


In [ ]:
# Group the data by 'director_name' and perform multiple aggregations
director_stats = data.groupby('director_name').agg(
    total_revenue=('revenue', 'sum'),
    average_vote=('vote_average', 'mean'),
    movie_count=('title', 'count')
)

# Sort the results by total_revenue in descending order
director_stats = director_stats.sort_values(by='total_revenue', ascending=False)

# Display the top directors based on these metrics
print("\n--- Director Performance Statistics (Sorted by Total Revenue) ---")
display(director_stats.head(10))


--- Director Performance Statistics (Sorted by Total Revenue) ---


,total_revenue,average_vote,movie_count
director_name,,,
Steven Spielberg,8676739164,6.953846,26
Peter Jackson,6498642820,7.333333,9
James Cameron,5883569439,7.328571,7
Michael Bay,5832524638,6.400000,12
Christopher Nolan,4227483234,7.800000,8
Chris Columbus,3725631503,6.590909,11
Robert Zemeckis,3590622002,6.915385,13
Tim Burton,3337418241,6.728571,14
Ridley Scott,3189557997,6.693750,16


### Quiz-2

# Question

Given a dataframe of employee rating info `(emp_id, monthly_rating)`.

How will you calculate the average rating of each employee?

# Choices
- [ ] `data.groupby('emp_id')`
- [ ] `data.groupby('monthly_rating')`
- [ ] `data.groupby('emp_id').mean()`
- [ ] `data.groupby('monthly_rating').mean()`

---

In [ ]:
data.head()

,Unnamed: 0,id_x,budget,popularity,revenue,title,vote_average,vote_count,year,month,day,director_name,gender,profit,is_successful
0,0,43597,237000000,150,2787965087,Avatar,7.2,11800,2009,Dec,Thursday,James Cameron,0,2550965087,True
1,1,43598,300000000,139,961000000,Pirates of the Caribbean: At World's End,6.9,4500,2007,May,Saturday,Gore Verbinski,0,661000000,True
2,2,43599,245000000,107,880674609,Spectre,6.3,4466,2015,Oct,Monday,Sam Mendes,0,635674609,True
3,3,43600,250000000,112,1084939099,The Dark Knight Rises,7.6,9106,2012,Jul,Monday,Christopher Nolan,0,834939099,True
4,5,43602,258000000,115,890871626,Spider-Man 3,5.9,3576,2007,May,Tuesday,Sam Raimi,0,632871626,True


In [ ]:
data[["director_name", "popularity", "revenue"]].groupby("director_name").mean()

,popularity,revenue
director_name,,
Adam McKay,30.333333,1.431794e+08
Adam Shankman,23.125000,1.091953e+08
Alejandro González Iñárritu,47.000000,1.463300e+08
Alex Proyas,53.200000,1.549125e+08
Alexander Payne,24.800000,8.204699e+07
...,...,...
Wes Craven,22.300000,7.647943e+07
Wolfgang Petersen,35.857143,2.307166e+08
Woody Allen,17.722222,3.449480e+07


In [ ]:
data.groupby("director_name").mean()

TypeError: agg function failed [how->mean,dtype->object]

In [ ]:
data.head()

,movies_id,budget,popularity,revenue,title,vote_average,vote_count,year,month,day,director_name,gender,profit,is_successful_method1,is_successful_method2
0,43597,237000000,150,2787965087,Avatar,7.2,11800,2009,Dec,Thursday,James Cameron,0,2550965087,True,True
1,43598,300000000,139,961000000,Pirates of the Caribbean: At World's End,6.9,4500,2007,May,Saturday,Gore Verbinski,0,661000000,True,True
2,43599,245000000,107,880674609,Spectre,6.3,4466,2015,Oct,Monday,Sam Mendes,0,635674609,True,True
3,43600,250000000,112,1084939099,The Dark Knight Rises,7.6,9106,2012,Jul,Monday,Christopher Nolan,0,834939099,True,True
4,43602,258000000,115,890871626,Spider-Man 3,5.9,3576,2007,May,Tuesday,Sam Raimi,0,632871626,True,True


### Quiz-3

# Question
For the following code, what will happen out of the given options?

```
data.groupby('director_name').mean()
```

# Choices
- [ ] It will give an error since we can't aggregate string columns like "title" using mean.
- [ ] It will show results only for columns where `mean()` aggregation is possible.
- [ ] It will show results for all the columns, with the non-applicable columns beind converted to their ASCII format for computation.

---

### Quiz-4

# Question
For each  year, how you will get the average, lowest, and highest value of movie ratings?

# Choices
- [ ] `data.groupby('year').agg({'vote_average': ['mean', 'min', 'max']})`
- [ ] `data.groupby['year'].agg({'vote_average': ['mean', 'min', 'max']})`
- [ ] `data.groupby('year').agg(('vote_average': ['mean', 'min', 'max']))`

---

## Group-based Filtering: Isolating Specific Groups

Group-based filtering allows us to select entire groups of data based on a condition that applies to the group as a whole, rather than filtering individual rows. This is different from `boolean masking`, which filters individual rows.

**Scenario:** Our film studio client wants to identify "high-budget directors." Let's define a "high-budget director" as any director who has directed **at least one movie with a budget greater than or equal to 100 million USD**.

To achieve this, we can follow a multi-step process or use the powerful `.filter()` method:

1.  **Step 1: Aggregate Max Budget per Director**
    First, we need to find the maximum budget for each director. We can **group** the `data` by `director_name` and then use the `max()` aggregation on the `budget` column.

1. We can **group** the data by director and use `max` of the budget as aggregator.

In [ ]:
# Group data by 'director_name' and find the maximum 'budget' for each director
# .reset_index() converts the grouped result back into a DataFrame
data_dir_budget = data.groupby("director_name")["budget"].max().reset_index()
# Display the first few rows of the resulting DataFrame
data_dir_budget.head()

,director_name,budget
0,Adam McKay,100000000
1,Adam Shankman,80000000
2,Alejandro González Iñárritu,135000000
3,Alex Proyas,140000000
4,Alexander Payne,30000000


2.  **Step 2: Identify High-Budget Director Names**
    Next, from our `data_dir_budget` DataFrame, we need to **filter** out the names of directors whose maximum movie budget was greater than or equal to 100 million USD. We can use boolean indexing for this.

In [ ]:
# Filter 'data_dir_budget' to get 'director_name' for directors whose maximum movie budget was greater than or equal to 100 million USD
names = data_dir_budget.loc[data_dir_budget["budget"] >= 100000000, "director_name"]

3.  **Step 3: Retrieve Movies by High-Budget Directors**
    Finally, with the list of `names` (our high-budget directors), we can filter our original `data` DataFrame to retrieve all movies directed by these specific individuals using the `.isin()` method.

In [ ]:
# Filter the original 'data' DataFrame to include all movies directed by the 'high-budget' directors identified in the previous step
# The .isin(names) method checks if each director's name in 'data' is present in our list of 'names'
data.loc[data['director_name'].isin(names)]

,id_x,budget,popularity,revenue,title,vote_average,vote_count,year,month,day,director_name,gender,profit,is_successful
0,43597,237000000,150,2787965087,Avatar,7.2,11800,2009,Dec,Thursday,James Cameron,0,2550965087,True
1,43598,300000000,139,961000000,Pirates of the Caribbean: At World's End,6.9,4500,2007,May,Saturday,Gore Verbinski,0,661000000,True
2,43599,245000000,107,880674609,Spectre,6.3,4466,2015,Oct,Monday,Sam Mendes,0,635674609,True
3,43600,250000000,112,1084939099,The Dark Knight Rises,7.6,9106,2012,Jul,Monday,Christopher Nolan,0,834939099,True
4,43602,258000000,115,890871626,Spider-Man 3,5.9,3576,2007,May,Tuesday,Sam Raimi,0,632871626,True
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1450,48267,400000,33,100000000,Mad Max,6.6,1213,1979,Apr,Thursday,George Miller,0,99600000,True
1451,48268,200000,13,4505922,Swingers,6.8,253,1996,Oct,Friday,Doug Liman,0,4305922,True
1452,48274,0,5,2611555,Three,6.3,31,2010,Dec,Thursday,Tom Tykwer,0,2611555,True
1458,48335,60000,27,3221152,Pi,7.1,586,1998,Jul,Friday,Darren Aronofsky,0,3161152,True


### Simplified Group Filtering with `.filter()`

The previous three-step process to filter groups is effective, but Pandas offers a more concise way to achieve the same result using the `.filter()` method directly on the `groupby` object. This method is specifically designed for group-based filtering.

**Question:** Can we filter groups in a single, elegant step using a `lambda` function with `.filter()`?

Yes! The `.filter()` method takes a function (often a `lambda` function for conciseness) that operates on each group. If the function returns `True` for a group, that entire group is included in the result; otherwise, it's excluded.

Let's try to achieve the same "high-budget director" filtering using `groupby().filter()`.

In [ ]:
# Use groupby().filter() to select entire groups (directors) where the maximum budget of their movies is >= 100
# The lambda function defines the condition for filtering each group (x represents a group's DataFrame)
data.groupby('director_name').filter(lambda x: x["budget"].max() >= 100)

,id_x,budget,popularity,revenue,title,vote_average,vote_count,year,month,day,director_name,gender,profit,is_successful
0,43597,237000000,150,2787965087,Avatar,7.2,11800,2009,Dec,Thursday,James Cameron,0,2550965087,True
1,43598,300000000,139,961000000,Pirates of the Caribbean: At World's End,6.9,4500,2007,May,Saturday,Gore Verbinski,0,661000000,True
2,43599,245000000,107,880674609,Spectre,6.3,4466,2015,Oct,Monday,Sam Mendes,0,635674609,True
3,43600,250000000,112,1084939099,The Dark Knight Rises,7.6,9106,2012,Jul,Monday,Christopher Nolan,0,834939099,True
4,43602,258000000,115,890871626,Spider-Man 3,5.9,3576,2007,May,Tuesday,Sam Raimi,0,632871626,True
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1460,48363,0,3,321952,The Last Waltz,7.9,64,1978,May,Monday,Martin Scorsese,0,321952,True
1461,48370,27000,19,3151130,Clerks,7.4,755,1994,Sep,Tuesday,Kevin Smith,0,3124130,True
1462,48375,0,7,0,Rampage,6.0,131,2009,Aug,Friday,Uwe Boll,0,0,False
1463,48376,0,3,0,Slacker,6.4,77,1990,Jul,Friday,Richard Linklater,0,0,False


### Understanding `groupby().filter()`

Notice what's happening here:
-   We first create a `groupby` object based on `director_name`.
-   Then, we use the `.filter()` method, passing it a `lambda` function. This function is applied to *each individual group*.
-   The `lambda x: x["budget"].max() >= 100` condition checks if the maximum budget within that specific director's group (`x`) is at least 100 million USD.
-   If the condition evaluates to `True` for a group, that entire group (all rows associated with that director) is retained in the resulting DataFrame. If `False`, the entire group is dropped.

This is known as **Group-Based Filtering** because we are making a decision about whether to keep or discard an entire group based on a characteristic of that group.

**Key Points:**
-   The `.filter()` method operates on **groups**, not individual rows.
-   The result of `groupby().filter()` is **not a `groupby` object**, but a regular **Pandas DataFrame** containing only the groups that satisfied the filtering criterion.

### Exercise 4: Filtering Groups

Let's put your group-filtering skills to the test!

**Task:**
1.  Filter the `data` DataFrame to include only those directors who have directed **more than 5 movies**.
2.  Display the `director_name` and the count of their movies for these filtered directors.

*Hint: You'll need to use `groupby().filter()` and a condition on the `size` of each group.*

In [ ]:
# 1. Filter the DataFrame to include only directors who have directed more than 5 movies
# The lambda function checks if the size of each group (number of movies) is > 5
experienced_directors = data.groupby('director_name').filter(lambda x: len(x) > 5)

# 2. Display the director_name and the count of their movies for these filtered directors
# Group the 'experienced_directors' DataFrame again and count movies per director
print("\n--- Directors with More Than 5 Movies ---")
display(experienced_directors['director_name'].value_counts().to_frame(name='movie_count'))


--- Directors with More Than 5 Movies ---


,movie_count
director_name,
Steven Spielberg,26
Martin Scorsese,19
Clint Eastwood,19
Woody Allen,18
Robert Rodriguez,16
...,...
Catherine Hardwicke,6
Michael Winterbottom,6
Atom Egoyan,6


---

## Group-based `apply()`: Performing Transformations within Groups

While `groupby()` with `.aggregate()` is great for summarizing data, and `groupby()` with `.filter()` is for selecting groups, `groupby()` with `.apply()` allows for more complex, arbitrary transformations on each group. This means you can run a custom function on each sub-DataFrame (group) independently and then combine the results.

**Scenario:** Our client wants to identify "risky movies." A movie is considered "risky" if its `budget` was higher than the *average `revenue` generated by that director's other movies*. This requires a calculation that uses an aggregate (average revenue) *within* each director's group, but then applies that to individual movies within the same group.

**How can we subtract the average `revenue` of a director from the `budget` column for each movie directed by that director, and then flag movies as 'risky'?**

We will define a custom function that takes a group (a sub-DataFrame for a single director) and calculates a new column based on group-specific aggregations.

In [ ]:
# Define a function to determine if a movie is 'risky' within a director's group
def func(x):
  # A movie is 'risky' if its budget is greater than or equal to the director's average revenue
  x["risky"] = x["budget"] - x["revenue"].mean() >= 0
  return x

# Group the data by 'director_name' and apply the 'func' function to each group
# 'group_keys=False' prevents the group key (director_name) from becoming part of the index
data_risky = data.groupby("director_name", group_keys=False).apply(func)

/tmp/ipykernel_40094/522996692.py:9: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  data_risky = data.groupby("director_name", group_keys=False).apply(func)


### Deconstructing `groupby().apply()`

Let's break down what we did with `groupby().apply()`:

1.  **Custom Function Definition (`func`):** We defined a function `func(x)` that takes a DataFrame `x` (which, when used with `groupby().apply()`, will be each sub-DataFrame corresponding to a director).
    - Inside `func`, we calculate the mean `revenue` for that specific director's group (`x["revenue"].mean()`).
    - Then, for *each movie* within that group, we compare its `budget` to the director's average `revenue`. If `budget - director_avg_revenue >= 0`, the movie is flagged as `True` (risky); otherwise, `False`.
    - The function then returns the modified group DataFrame `x`.

2.  **`groupby("director_name", group_keys=False).apply(func)`:**
    - We group the `data` DataFrame by `director_name`.
    - `group_keys=False`: This parameter prevents Pandas from adding the group key (`director_name` in this case) as an index level in the output. Setting it to `False` keeps the output DataFrame structure cleaner. (Note: The default behavior of `group_keys` might change in future Pandas versions).
    - The `.apply(func)` then takes our `func` and applies it to *each* `director_name` group. Pandas passes each sub-DataFrame `x` to our `func`.

**What was the overall process here?**
-   We defined a custom function (`func`) that encapsulates our specific business logic.
-   We grouped the `data` by `director_name`.
-   We used the `.apply()` method with our custom function on the grouped data, allowing us to perform a calculation that combines both individual row data (`budget`) with group-level aggregates (mean `revenue`).

Now, let's see if there are any movies that our analysis has flagged as "risky" based on this criterion.

In [ ]:
# Filter the 'data_risky' DataFrame to show only the movies flagged as 'risky' (where the 'risky' column is True)
data_risky.loc[data_risky["risky"]]

,id_x,budget,popularity,revenue,title,vote_average,vote_count,year,month,day,director_name,gender,profit,is_successful,risky
7,43608,200000000,107,586090727,Quantum of Solace,6.1,2965,2008,Oct,Thursday,Marc Forster,0,386090727,True,True
12,43614,380000000,135,1045713802,Pirates of the Caribbean: On Stranger Tides,6.4,4948,2011,May,Saturday,Rob Marshall,0,665713802,True,True
15,43618,200000000,37,310669540,Robin Hood,6.2,1398,2010,May,Wednesday,Ridley Scott,0,110669540,True,True
20,43624,209000000,64,303025485,Battleship,5.5,2114,2012,Apr,Wednesday,Peter Berg,0,94025485,True,True
24,43630,210000000,3,459359555,X-Men: The Last Stand,6.3,3525,2006,May,Wednesday,Brett Ratner,0,249359555,True,True
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1347,47224,5000000,7,3263585,The Sweet Hereafter,6.8,103,1997,May,Wednesday,Atom Egoyan,0,-1736415,False,True
1349,47229,5000000,3,4842699,90 Minutes in Heaven,5.4,40,2015,Sep,Friday,Michael Polish,0,-157301,False,True
1351,47233,5000000,6,0,Light Sleeper,5.7,15,1992,Aug,Friday,Paul Schrader,1,-5000000,False,True
1356,47263,15000000,10,0,Dying of the Light,4.5,118,2014,Dec,Thursday,Paul Schrader,1,-15000000,False,True


### Exercise 5: Analyzing Risky Movies

Based on our `risky` flag, let's dive deeper into these potentially problematic productions.

**Task:**
1.  Filter the `data_risky` DataFrame to show only the movies that are flagged as `risky` (`risky == True`).
2.  For these risky movies, display the `director_name`, `title`, `budget`, `revenue`, `profit`, and the `risky` flag.
3.  Calculate the total number of risky movies identified.

*This exercise helps reinforce filtering and interpretation of new features created via `apply()`.*

In [ ]:
# 1. Filter the 'data_risky' DataFrame to show only the movies flagged as 'risky'
risky_movies = data_risky[data_risky['risky'] == True]

# 2. Display the requested columns for these risky movies
print("\n--- Risky Movies Analysis ---")
display(risky_movies[['director_name', 'title', 'budget', 'revenue', 'profit', 'risky']].head())

# 3. Calculate and display the total number of risky movies
total_risky_movies = len(risky_movies)
print(f"\nTotal number of risky movies identified: {total_risky_movies}")


--- Risky Movies Analysis ---


,director_name,title,budget,revenue,profit,risky
7,Marc Forster,Quantum of Solace,200000000,586090727,386090727,True
12,Rob Marshall,Pirates of the Caribbean: On Stranger Tides,380000000,1045713802,665713802,True
15,Ridley Scott,Robin Hood,200000000,310669540,110669540,True
20,Peter Berg,Battleship,209000000,303025485,94025485,True
24,Brett Ratner,X-Men: The Last Stand,210000000,459359555,249359555,True



Total number of risky movies identified: 131


### Quiz-5

# Question

Given the following DataFrame:

```python=
import pandas as pd
data = {'Group': ['G1', 'G2', 'G2', 'G3', 'G3', 'G3'],
        'Points': [2, 4, 6, 8, 10, 12]}
df = pd.DataFrame(data)
```

What will be output of the code given below?

```python=
result = df.groupby('Group')['Points'].apply(lambda x: x.sum())
print(result)
```

A.
```
Group
G1     2
G2    10
G3    30
```

B.
```
Group
G1     2
G2     4
G3     8
```

C.
```
Group
G1     2
G2     6
G3    12
```

D.
```
Group
G1     2
G2    20
G3    60
```

# Choices
- [ ] A
- [ ] B
- [ ] C
- [ ] D

---

### Practice Coding Question

- [Group the Data](https://www.scaler.com/hire/test/problem/23273/)

### Assignment Questions

- [Is this the end ?](https://www.scaler.com/hire/test/problem/28383/)
- [Apply pandas](https://www.scaler.com/hire/test/problem/14715/)
- [Mean Income](https://www.scaler.com/hire/test/problem/28702/)
- [On your marks](https://www.scaler.com/hire/test/problem/28720/)
- [Categorise ages](https://www.scaler.com/hire/test/problem/23233/)
- [Haven't followed the instructions?](https://www.scaler.com/hire/test/problem/19155/)
- [The island of slice](https://www.scaler.com/hire/test/problem/28060/)

### Homework Questions

- [StockCode in UK](https://www.scaler.com/hire/test/problem/15176/)
- [Most selling Product](https://www.scaler.com/hire/test/problem/15169/)
- [Group the Data](https://www.scaler.com/hire/test/problem/23273/)
- [Highest Amount Purchase](https://www.scaler.com/hire/test/problem/15170/)
- [Student Filtering](https://www.scaler.com/hire/test/problem/23261/)